Mathlib/RingTheory/Coalgebra.lean

```diff
@@ -0,0 +1,116 @@
+/-
+Copyright (c) 2023 Ali Ramsey. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Ali Ramsey
+-/
+import Mathlib.RingTheory.TensorProduct
+
+/-!
+# Definition and example of a coalgebra
+-/
```

Comment:
Ideally we'd have a longer docstring; a good start would be
```suggestion
/-!
# Coalgebras

In this file we define `Coalgebra`, and provide instances for:

* Commutative rings: `CommRing.toCoalgebra`
* Finitely supported functions: `Finsupp.instCoalgebra`

## References

* Some ncatlab / wikipedia / bourbaki / ... reference of your choosing
-/
```
I think @adamtopaz can probably suggest something a little more in depth

Reply:
I agree that a better module docstring would be more desirable -- this is the first thing people see in the docs when they navigate to the coalgebra page so it should reflect the contents of the file.